ext/ManoptJuMPExt.jl

```diff
@@ -0,0 +1,387 @@
+module ManoptJuMPExt
+
+using Manopt
+using LinearAlgebra
+using JuMP: JuMP
+const MOI = JuMP.MOI
+using ManifoldsBase
+using ManifoldDiff
+
+function __init__()
+    # So that the user can use the convenient `Manopt.Optimizer`
+    setglobal!(Manopt, :Optimizer, Optimizer)
+    return nothing
+end
+
+"""
+    struct VectorizedManifold{M} <: MOI.AbstractVectorSet
+        manifold::M
+    end
+
+Representation of points of `manifold` as a vector of `R^n` where `n` is
+`MOI.dimension(VectorizedManifold(manifold))`.
+"""
+struct VectorizedManifold{M<:ManifoldsBase.AbstractManifold} <: MOI.AbstractVectorSet
+    manifold::M
+end
+
+"""
+    MOI.dimension(set::VectorizedManifold)
+
+Return the dimension of the vectorized Manifold. As the MOI variables are real
+if the dimension is `n`, it means that the vectorization is a subset of `R^n`.
+"""
+function MOI.dimension(set::VectorizedManifold)
+    return prod(ManifoldsBase.representation_size(set.manifold))
+end
+
+"""
+    Manopt.Optimizer()
+
+Creates a new optimizer object.
+
+The minimization of a function `f(X)` of an an array `X[1:n1,1:n2,...]`
+over a manifold `M` starting at `X0`, can be modeled as follows:
+```julia
+using JuMP
+@variable(model, X[i1=1:n1,i2=1:n2,...] in M, start = X0[i1,i2,...])
+@objective(model, Min, f(X))
+```
+The optimizer assumes that `M` has a `Array` shape described
+by `ManifoldsBase.representation_size`.
+"""
+mutable struct Optimizer <: MOI.AbstractOptimizer
+    # Manifold in which all the decision variables leave
+    manifold::Union{Nothing,ManifoldsBase.AbstractManifold}
+    # Description of the problem in Manopt
+    problem::Union{Nothing,Manopt.AbstractManoptProblem}
+    # State of the optimizer
+    state::Union{Nothing,Manopt.AbstractManoptSolverState}
+    # Starting value for each variable
+    variable_primal_start::Vector{Union{Nothing,Float64}}
+    # Sense of the optimization, e.g., min, max or no objective
+    sense::MOI.OptimizationSense
+    # Model used to compute gradient of the objective function with AD
+    nlp_model::MOI.Nonlinear.Model
+    # Solver parameters set with `MOI.RawOptimizerAttribute`
+    options::Dict{String,Any}
+    function Optimizer()
+        return new(
+            nothing,
+            nothing,
+            nothing,
+            Union{Nothing,Float64}[],
+            MOI.FEASIBILITY_SENSE,
+            MOI.Nonlinear.Model(),
+            Dict{String,Any}(),
+        )
+    end
+end
+
+MOI.get(::Optimizer, ::MOI.SolverVersion) = "0.4.37"
+
+function MOI.is_empty(model::Optimizer)
+    # TODO replace `isnothing(model.nlp_model.objective)`
+    #      by `MOI.is_empty` once the following is fixed
+    #      https://github.com/jump-dev/MathOptInterface.jl/issues/2302
+    return isnothing(model.manifold) &&
+           isempty(model.variable_primal_start) &&
+           isnothing(model.nlp_model.objective) &&
+           model.sense == MOI.FEASIBILITY_SENSE
+end
+
+function MOI.empty!(model::Optimizer)
```

Comment:
does this need to be documented maybe? Also just seeing the signature ere, I would not see, that this would only be the manopt one – for readability one could withe `Manopt.` upfront the type to help the reader (code-wise no difference, I know).

Reply:
You mean everywhere ? Isn't that enough that it's in the `ManoptJuMPExt` scope ? Our approach to this is to only import the dependencies so that anytime there is no prefix, you know it's something from the current module. Otherwise, you have the explicit package where it's coming from.

Reply:
We can add the prefix in the docstring though

Reply:
Hm, to be honest I am not yet 100% sure here what I want myself.

Imagine that at some point I have the great idea to combine my `problem` and `state` into an `optimizer` (one might argue – maybe go for `solver` then, sure, but then my point is not so easy to get across).
Then your name/struct would conflict with an existing `Optimizer` struct in Manopt. 

On a more philosophical level, `Manopt.Optimizer` sounds like the struct I just envisioned in the last section, but it actually is not that, but the indicator for JuMP (and also that only if JuMP is loaded) – I am not yet sure an indicator/struct for what, but that is another problem.
So while it should be documented, I am also not so much in favour of the name here, since it might cause confusion – and you are basically “pirating” in “my” (Manopt.jls) namespace.

Reply:
Then I agree, when you read the docstring, you are not sure in which module you are so the context is less clear :)